mlflow/spark.py

```diff
@@ -34,12 +36,67 @@ def log_model(spark_model, artifact_path, conda_env=None, jars=None):
     :param conda_env: Path to a Conda environment file. If provided, defines environment for the
     model. At minimum, it should specify python, pyspark, and mlflow with appropriate versions.
     :param jars: List of JARs needed by the model.
+    :param dfs_tmpdir: Temporary directory path on DFS or local FS if running in local mode. The
+    model will be writen in this destination and than copied to requested local path. This is
+    necessary as Spark ML models read / write from / to DFS if running on a cluster. All temporary
+    files created on the DFS will be removed if this operation completes successfully.
     """
     return Model.log(artifact_path=artifact_path, flavor=mlflow.spark, spark_model=spark_model,
-                     jars=jars, conda_env=conda_env)
+                     jars=jars, conda_env=conda_env, dfs_tmpdir=dfs_tmpdir)
+
+
+def _tmp_path(dfs_tmp):
+    import uuid
+    return os.path.join(dfs_tmp, str(uuid.uuid4()))
+
+
+class _HadoopFileSystem:
+    """
+    Interface to org.apache.hadoop.fs.FileSystem.
+
+    Spark ML models expect to read from and write to Hadoop FileSystem when running on a cluster.
+    Sine MLflow works on local directories, we need this interface to copy teh files between
+    the current DFS and local dir.
+    """
+    _filesystem = None
+    _conf = None
+
+    @classmethod
+    def _jvm(cls):
+        return SparkContext._gateway.jvm
+
+    @classmethod
+    def _fs(cls):
+        if not cls._filesystem:
+            sc = SparkContext.getOrCreate()
+            cls._conf = sc._jsc.hadoopConfiguration()
+            cls._filesystem = cls._jvm().org.apache.hadoop.fs.FileSystem.get(cls._conf)
+        return cls._filesystem
+
+    @classmethod
+    def _local_path(cls, path):
+        return cls._jvm().org.apache.hadoop.fs.Path(os.path.abspath(path))
+
+    @classmethod
+    def _remote_path(cls, path):
+        return cls._fs().makeQualified(cls._jvm().org.apache.hadoop.fs.Path(path))
```

Comment:
Do we want to keep the `makeQualified` piece? Is it to support remote paths that are relative?

Reply:
Yes, I left it in there because that's how I tested it. I can try to leave it and see if it still works.